src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -85,26 +84,63 @@ public Task RunAsync(ActionRunStage stage)
                 actionRunner.Stage = stage;
                 actionRunner.Condition = aStep.Condition;
 
-                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, location, Environment);
-
+                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, Environment);
                 InitializeScope(step);
 
-                location++;
+                // Add all steps to the Composite StepsRunner
+                // Follows similar logic to how JobRunner invokes the StepsRunner for job steps!
+                compositeSteps.Add(step);
             }
 
-            // Create a step that handles all the composite action steps' outputs
-            Pipelines.ActionStep cleanOutputsStep = new Pipelines.ActionStep();
-            cleanOutputsStep.ContextName = ExecutionContext.ContextName;
-            // Use the same reference type as our composite steps.
-            cleanOutputsStep.Reference = Action;
+            try
+            {
+                // This is where we run each step.
+                await RunStepsAsync(compositeSteps);
 
-            var actionRunner2 = HostContext.CreateService<IActionRunner>();
-            actionRunner2.Action = cleanOutputsStep;
-            actionRunner2.Stage = ActionRunStage.Main;
-            actionRunner2.Condition = "always()";
-            ExecutionContext.RegisterNestedStep(actionRunner2, inputsData, location, Environment, true);
+                // This is where we set the outputs.
+                // All steps have the same scope for their ExecutionContext/
+                // So we can use one of them so that we can get the right "outputs" attribute.
+                HandleOutput(compositeSteps?[0].ExecutionContext);
```

Comment:
compositeSteps is never null

Reply:
should we check if compositeSteps.Count \> 0 (guard against index out of range)?

Reply:
Should handle-output get it's own child node instead? I wonder what unintentional side effects this will have by re-using the context for a previous step that already ran.